### DISCUSSION TITLE
P. generosa functional gene annotations

### DISCUSSION THREAD

**Discussion by ocattau** [RobertsLab/resources#1324] [Q&A]
After I have my count data for a single library for P. generosa, I need to annotate my table with gene functions. Do we already have a Blast for v5 of the transcriptome? If so where would I find it?

**Comment by kubu4** [comment]
On Mox: `/gscratch/srlab/blastdbs/ncbi-sp-v5_20210224/`

**Reply by ocattau** [reply]
Okay new question for Mox newbs, how to I download this file in a way I can use R to analyze it?

**Comment by kubu4** [comment]
Which file?

**Reply by ocattau** [reply]
ncbi-sp-v5_20210224/swissprot.tar.gz

I know I need to use rsync but where can I upload it to? Does it have to be gannet?

**Comment by kubu4** [comment]
You don't want/need that file. You'll want to run BLAST on Mox, which would utilize that BLAST database.

**Reply by ocattau** [reply]
Ah That is true, okay so there isn't full transciptome in Mox then. I have to blast again.

**Comment by kubu4** [comment]
You'll want to `rsync` your query file (e. g. a FastA file) to your directory on `/gscratch/srlab/` (you might need to create your directory) or `/gscratch/ scrubbed/`.

Then, run BLAST and use that BLAST database listed above.

**Reply by kubu4** [reply]
Sorry, I misunderstood the original question. My brain filled in "database" in your original question.

Anyway, check the Lab Handbook to see if an annotation file exists for the geoduck transcriptome.

**Reply by ocattau** [reply]
And when I do my blasting, do I just use the build node?

**Reply by sr320** [reply]
nope - execute node.. done via slurm script

https://robertslab.github.io/resources/mox_Node-Types/

**Reply by sr320** [reply]
@ocattau This might be helpful https://youtu.be/1RiWiRx-B9I

**Reply by ocattau** [reply]
Okay new problem, Slurm Job failed and I got Exit code 1. Can you look at my slurm script to make sure it looks okay? 

```{bash}
#!/bin/bash
## 10.29.21 blast p.generosa on MOX
#SBATCH --job-name=blasting_p.generosa_10.29.21
## Allocation Definition
#SBATCH --account=srlab
#SBATCH --partition=srlab
## Nodes
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=0-12:00:00
## Memory per node
#SBATCH --mem=500G
##turn on e-mail notification
#SBATCH --mail-type=ALL
#SBATCH --mail-user=ocattau@uw.edu
## Specify the working directory for this job
#SBATCH --chdir=/gscratch/srlab/ocattau/102921-p.generosa

source /gscratch/srlab/programs/scripts/slurm-p.generosa.sh

/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/blastx \
-query /gscratch/srlab/ocattau/p.generosa/Panopea-generosa-v1.0.fa \
-db /gscratch/srlab/blastdbs/uniprot_sprot_20200123/uniprot_sprot.fasta  \
-out /gscratch/srlab/ocattau/p.generosa/102921-p.generosa/Panopea-generosa-uniprot_blastx.tab \
-evalue 1E-5 \
-num_threads 10 \
-max_target_seqs 1 \
-max_hsps 1 \
-outfmt 6
```

**Reply by sr320** [reply]
You need to look at slurm out file to see error

**Reply by ocattau** [reply]
Is this because I didn't set the account to ocattau?

```{bash}
Command line argument error: Argument "out". File is not accessible:  `/gscratch/srlab/ocattau/p.generosa/102921-p.generosa/Panopea-generosa-uniprot_blastx.tab'
```

**Reply by kubu4** [reply]
The reason your job died is actually due to this error (this is the line above the error you posted):

`/gscratch/srlab/programs/scripts/slurm-p.generosa.sh: No such file or directory`

And, I think this is caused by this line in your SLURM script:

`source /gscratch/srlab/programs/scripts/slurm-p.generosa.sh`

You can delete this line and I think your script will run just fine.

**Reply by kubu4** [reply]
Also, for organizational purposes, you'll usually (always?) want to put the SLURM script in your working directory (i.e. the directory specified on this line of the SLURM header):

`#SBATCH --chdir=`

**Reply by ocattau** [reply]
My new problem is my fasta file. I have run and re-run my slurm with various iterations of my code above and I consistently get this message. Am I using the wrong fasta file? 
```{bash}
BLAST Database error: No alias or index file found for protein database [/gscratch/srlab/ocattau/p.generosa/uniprot_sprot.fasta] in search path [/gscratch/srlab/ocattau/p.generosa::]
```

My new slurm script
```{bash}
#!/bin/bash
## 10.29.21 blast p.generosa on MOX
#SBATCH --job-name=blasting_p.generosa_10.29.21
## Allocation Definition
#SBATCH --account=srlab
#SBATCH --partition=srlab
## Nodes
#SBATCH --nodes=1
## Walltime (days-hours:minutes:seconds format)
#SBATCH --time=00-12:00:00
## Memory per node
#SBATCH --mem=500G
##turn on e-mail notification
#SBATCH --mail-type=ALL
#SBATCH --mail-user=ocattau@uw.edu
## Specify the working directory for this job
#SBATCH --chdir=


/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/blastx \
-query /gscratch/srlab/ocattau/p.generosa/Panopea-generosa-v1.0.fa \
-db /gscratch/srlab/ocattau/p.generosa/uniprot_sprot.fasta \
-out /gscratch/srlab/ocattau/p.generosa/Panopea-generosa-uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 30 \
-max_target_seqs 1 \
-max_hsps 1 \
-outfmt 6

```

**Reply by sr320** [reply]
you do not use a fasta file.. needs to be a blast database....

**Reply by kubu4** [reply]
You should change your `-db` line to:

`-db /gscratch/srlab/blastdbs/ncbi-sp-v5_20210224/swissprot.pdb`

**Reply by kubu4** [reply]
Actually, you might need to leave off the `.pdb` at the end of that line... Can't remember, ATM.

**Reply by kubu4** [reply]
Yep, the line needs to be:

`-db /gscratch/srlab/blastdbs/ncbi-sp-v5_20210224/swissprot`

Sorry for the confusion!!!

**Reply by ocattau** [reply]
The video you sent uses the fasta so I got confused

**Reply by sr320** [reply]
actually it was not a fasta file... just had the fasta extension... it was a blast database.

**Reply by kubu4** [reply]
> The video you sent uses the fasta so I got confused

Sorry, that's an unfortunate/confusing naming thing. If you look in that directory, you'll see what the files are called. Here's an example:

```shell
ls -l /gscratch/srlab/blastdbs/UniProtKB_20190109
total 574976
-rw-r--r-- 1 samwhite hyak-srlab       263 Jan  9  2019 readme.txt
-rw-r--r-- 1 samwhite hyak-srlab 275880518 Jan  9  2019 uniprot_sprot.fasta
-rw-r--r-- 1 samwhite hyak-srlab 106931528 Jan  9  2019 uniprot_sprot.fasta.phr
-rw-r--r-- 1 samwhite hyak-srlab   4471264 Jan  9  2019 uniprot_sprot.fasta.pin
-rw-r--r-- 1 samwhite hyak-srlab 201309527 Jan  9  2019 uniprot_sprot.fasta.psq
```

The `-db` argument requires you to specify the basename of the database. So, in this instance, you would specify `uniprot_sprot.fasta`. Although there _is_ an actual FastA file (this is all very easy to follow, isn't it? Ugh.), the `-db` argument is looking for the three files ending in `.phr`, `.pin` and `.psq`. So, it expects the basename for all three of those files, which happens to be: `uniprot_sprot.fasta`.

Related to this, you should try to use one of the BLAST DBs we have available. I'd recommend against moving/copying something to your own directory. The primary reason for this recommendation is to improve reproducibility. If you use a BLAST DB that's established in one of the shared locations, it will have documentation about when/how/who generated it _and_ other people can use the same DB to help test out reproducibility. Plus, using one of the established DBs will help us troubleshoot if things go wrong.

EDITED: Acknowledge confusion of existence of a FastA file in the BLAST DB directory example above.

**Reply by sr320** [reply]
> it will have documentation about when/how/who generated it

we just looked at the folder and did not find a readme, so it was not clear to us what all the databases were.

**Reply by kubu4** [reply]
Which folder? I'm not even sure what database is trying to be utilized.

**Reply by sr320** [reply]
I believe we were looking at '/gscratch/srlab/blastdbs'

**Reply by kubu4** [reply]
There are a bunch of DBs directories in there. There are readmes present in each of the BLAST DBs which describe how/who/when/etc.

E.g.

```
cat /gscratch/srlab/blastdbs/uniprot_sprot_20200123/readme.txt


Directory containing files for BLASTing against the Uniprot SwissProt database.

---

FILES

- uniprot_sprot.fasta: Uniprot SwissProt protein FastA file. Downloaded 20200123.

- uniprot_sprot.dmnd: DIAMOND Uniprot (Reviewed) BLAST database generated with the following command:

makedb --in uniprot_sprot.fasta --threads 27 --db uniprot_sprot --taxonmap ../ncbi-nr-20190925/prot.accession2taxid --taxonnodes ../ncbi-nr-20190925/nodes.dmp
```

**Comment by ocattau** [comment]
Okay starting now thread in this discussion but if I want to blast a scaffold Panopea-generosa-v1.0.fa against the geoduck db online at https://owl.fish.washington.edu/halfshell/genomic-databank/Pgenerosa_transcriptome_v5.fasta that I dowloaded from the command line into my folder in Mox, so my code looks like this:

```{bash}
/gscratch/srlab/programs/ncbi-blast-2.8.1+/bin/blastx \
-query /gscratch/srlab/ocattau/p.generosa/Panopea-generosa-v1.0.fa \
-db /gscratch/srlab/ocattau/p.generosa/Pgenerosa_transcriptome_v5.fasta 
-out /gscratch/srlab/ocattau/p.generosa/Panopea-generosa-uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 30 \
-max_target_seqs 1 \
-max_hsps 1 \
-outfmt 6
```

Again I am getting the same error message that -db /gscratch/srlab/ocattau/p.generosa/Pgenerosa_transcriptome_v5.fasta  is not a database or index. 

```{bash}
# the code for Pgenerosa_transcriptome_v5.fasta 
TRINITY_DN88518_c0_g1_i1 len=571 path=[1:0-44 2:45-525 4:526-570]
CAATAACAGCTTATATTTGGAAATGAACAGATTAACATAAGCTGGATGGGAGTCCATTGATTGGGGTGTTGATTTCTCATGTGAAAGAAATGGAATTGAGACCAATTCATAGCTTAGTTTACAAGTTTAGGGTCTGGGGTTATAAAACTTGGTTTGTAATTGATAAAATTTTAGCCATACTTGTTTGTAGTTGATAAAGTTACAGTCCTTGTAGTTATGGTAACCATATCCTCTGTATAAATAGACATGGTAAATGTTACATTATATCTGACTGTTGTTATAGTGGTGTCAGCAGGTCGCAGAAAGAAGGTAAGCTTGGCTGTGATTTGCTTTAGAAAAAATAATTAAGTGTGATTGATTGATTGATCTATTTATTGATTGGATTTGATTCTGATGAATTTCCATAAAAAAAATAAACTTGGAGTAGTAGTGTTTTGAGTAATAAATATTTTACATTTTTTCAACTTCAGTTCTGTTATCTATTGTTTTTCCTTTTTTTTCTGTAAATAATTATATACTTAGATTACGACTTTGTGAATTCAAATTTATTTAAATAATTGTGAATCAGCAC
```

Is this not a database?

**Reply by kubu4** [reply]
> Is this not a database?

It is not. It is a FastA file.

If you want to BLAST against that, you need to create a BLAST database using the `makeblastdb` command (it's part of the BLAST suite of programs).

**Reply by sr320** [reply]
more importantly you DO NOT want to blast a scaffold Panopea-generosa-v1.0.fa against the geoduck db

**Reply by ocattau** [reply]
So we are back to -db=swiss-prot? When I ran the code yesterday it timed out after 12 hours so I feel like that was incorrect also. What db should I be using then?

**Reply by kubu4** [reply]
> it timed out after 12 hours so I feel like that was incorrect also

Why do you think so?

BLAST-ing will take a long time. 

Can you please post your SLURM script which was running successfully and timed out?
